### PR TITLE
Code Quality: Refactored code for security settings

### PR DESCRIPTION
**Pull request by onein528**
### Description

I'm working on refactoring security codes.

### Motivation and Context

The storage object security codes were noticeably unmaintable; it was way too difficult to understand what the class represents from the class name.

- Related #4180

### Tasks

- [x] Force to be one class per file.
- [x] Rename all security related classes with more Win32API-like naming convention.
- [x] Rename variables and methods too.
- [x] Add xml comments to all classes to describe that what the class represents.
- [x] Integrate classes that don't need to be separated.

### Changes

- Rename `Files.App.Filesystem.Permissions` with `Files.App.Filesystem.Security`.
- Rename the classes in `Files.App.Filesystem.Security` to improve ease of understanding what the class represents.
- Refactor the classes in `Files.App.Filesystem.Security`.
- Add xml comments in all classes in `Files.App.Filesystem.Security`.
- Refactor `Files.App.ViewModels.Properties.SecurityViewModel`.
- Refactor code-behinds of `Files.App.Views.Properties.SecurityPage` and `Files.App.Views.Properties.SecurityAdvancedPage`.

### My changes validation

**Preferable terms**
Name|Description|Rationale|Instead of
---|---|---|---
|AccessControlList|Represents a storage object's access control list|ACL<br>Used in Win32API|*None*
|AccessControlEntry|Represents an entry of access control list|ACE<br>ACCESS_ALLOWED_ACE<br>ACCESS_DENIED_ACE<br>Used in Win32API|*None*
|AccessMask|Represents access control flags(enum) of an ACE|ACCESS_MASK<br>Used in Win32API|*None*
|Principal|Represents an owner of ACL or ACE|Used in File Explorer UI of security advanced page|*None*
|Allow[ed]|*Literally*|Used in Win32API|Grant[ed\|s]
|Deny[ed]|*Literally*|Used in Win32API|*None*

**Unpreferable terms**
Name|Description|Rationale|Instead,
---|---|---|---
|Grant[ed\|s]|*Literally*|This term is not used in Win32API|Allow[ed]
|Permission[s]|*Literally*|This term is not used in Win32API|Security or *something else*

**Classes**
Name|Previous name|Rationale
---|---|---
`App.Filesystem.Permissions` | `App.Filesystem.Security` | that was unpreferable term. `Security` is used in File Explorer UI
||
`AccessControlEntry` | `FileSystemAccessRule` | Used as `ACE` of Win32API for SecurityPage UI
*Removed* | `FileSystemAccessRuleForUI` | Used as `ACE` of Win32API for SecurityAdvancedPage UI
*Removed* | `FileSystemAccessRule` | No references
`AccessControlEntryPrimitiveMapping` | `FileSystemAccessRule2` | *Apparently*
`AccessControlList` | `FilePermissions` | Used as `ACL` of Win32API
*Removed* | `FilePermissionsManager` | This class was the same as `AccessControlList`
`AccessControlType` | `AccessControlType`  | *None*
`AccessMaskFlags` | `FileSystemRights` | `FileSystemRights` is not familiar for Win32API devs. Equivalent to ACCESS_MASK of Win32API.
`AccessMaskItem` | `GrantedPermission` | Represents details of one flag of `AccessMaskFlags`.
`InheritanceFlags` | `InheritanceFlags` | *None*
`Principal` | `UserGroup` | This class can be either User or Group(`UserGroup` isn't accurate). `Principal` is used in File Explorer UI
`PrincipalType` | `SecurityType` | *Apparently*
`PropagationFlags` | `PropagationFlags` | *None*

### Validation

- [x] Built and ran the app
- [x] Tested the changes for accessibility

### Test

- [x] Display ACL and all ACEs
- [x] Add an ACE
- [x] Remove an ACE

### Screenshots

image

### CODE REVIEW

File: src/Files.App/Dialogs/FilesystemOperationDialog.xaml [hunk 1]
```diff
@@ -105,15 +105,15 @@
...
 				<ComboBox
 					x:Name="ConflictOptions"
 					Grid.Column="2"
 					Width="200"
 					HorizontalAlignment="Right"
 					VerticalAlignment="Center"
-					Visibility="{x:Bind IsConflict, Mode=OneWay, Converter={StaticResource BoolToVisibilityConverter}}"
-					SelectedIndex="{x:Bind ConflictResolveOption, UpdateSourceTrigger=PropertyChanged, Mode=TwoWay, Converter={StaticResource ConflictResolveOptionToIndexConverter}}">
+					SelectedIndex="{x:Bind ConflictResolveOption, UpdateSourceTrigger=PropertyChanged, Mode=TwoWay, Converter={StaticResource ConflictResolveOptionToIndexConverter}}"
+					Visibility="{x:Bind IsConflict, Mode=OneWay, Converter={StaticResource BoolToVisibilityConverter}}">
```
Comment (yaira2) [inline src/Files.App/Dialogs/FilesystemOperationDialog.xaml:116]: These changes look unrelated, let's try to limit PRs to their purpose. Makes it easier to review and to look back later on.

File: src/Files.App/Filesystem/Permissions/AccessControlEntryAdvanced.cs [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+				accessMaskStrings.Add("SecurityModifyLabel/Text".GetLocalizedResource());
+			}
+			else if (GrantsReadAndExecute)
+			{
+				accessMaskStrings.Add("SecurityReadAndExecuteLabel/Text".GetLocalizedResource());
+			}
+			else if (GrantsRead)
+			{
+				accessMaskStrings.Add("SecurityReadLabel/Text".GetLocalizedResource());
+			}
```
Comment (yaira2) [inline src/Files.App/Filesystem/Permissions/AccessControlEntryAdvanced.cs:423]: Switch case?

File: src/Files.App/Views/SettingsPages/Folders.xaml [hunk 3]
```diff
@@ -140,8 +140,8 @@
 						<local:SettingsBlockControl Title="{helpers:ResourceString Name=GroupInDescendingOrder}" HorizontalAlignment="Stretch">
 							<ToggleSwitch
 								AutomationProperties.Name="{helpers:ResourceString Name=GroupInDescendingOrder}"
-								IsOn="{x:Bind ViewModel.GroupInDescendingOrder, Mode=TwoWay}"
 								IsEnabled="{x:Bind ViewModel.isDefaultGrouped, Mode=OneWay}"
+								IsOn="{x:Bind ViewModel.GroupInDescendingOrder, Mode=TwoWay}"
```
Comment (yaira2) [inline src/Files.App/Views/SettingsPages/Folders.xaml:144]: Unrelated change

File: src/Files.App/UserControls/MultitaskingControl/HorizontalMultitaskingControl.xaml [hunk 4]
```diff
@@ -608,8 +608,8 @@
 						Drop="TabViewItem_Drop"
 						Header="{x:Bind Header, Mode=OneWay}"
 						IconSource="{x:Bind IconSource, Mode=OneWay}"
-						Style="{StaticResource Local.TabViewItem}"
 						Loaded="TabViewItem_Loaded"
+						Style="{StaticResource Local.TabViewItem}"
```
Comment (yaira2) [inline src/Files.App/UserControls/MultitaskingControl/HorizontalMultitaskingControl.xaml:612]: Please revert

**Comment by onein528** [comment]
Microsoft.Windows.CsWin32 was too dificult to understand how to handle strings. They don’t accept StringBuilder. Instead, they uses PWSTR. But I couldn’t how to assign.

File: src/Files.App/Filesystem/Permissions/AccessControlEntryAdvanced.cs (hunk 2 above)
Comment (onein528) [reply src/Files.App/Filesystem/Permissions/AccessControlEntryAdvanced.cs]: No, I cannot. Because those if-else-if uses different values.

File: src/Files.App/ViewModels/Properties/SecurityViewModel.cs [hunk 5]
```diff
@@ -139,18 +119,43 @@ public string DisableInheritanceOption
...
+		public RelayCommand RemoveRulesForUserCommand { get; set; }
+		public RelayCommand AddAccessRuleCommand { get; set; }
+		public RelayCommand RemoveAccessRuleCommand { get; set; }
+		public RelayCommand DisableInheritanceCommand { get; set; }
+		public RelayCommand<string> SetDisableInheritanceOptionCommand { get; set; }
+		public RelayCommand ReplaceChildPermissionsCommand { get; set; }
+		#endregion
+
+		#region Methods
+		private void InitializeCommands()
```
Comment (ferrariofilippo) [inline src/Files.App/ViewModels/Properties/SecurityViewModel.cs:133]: Can you remove region?

File: src/Files.App/ViewModels/Properties/SecurityViewModel.cs [hunk 6]
```diff
@@ -139,18 +119,43 @@ public string DisableInheritanceOption
...
+		}
+
+		private void DisableInheritance()
+		{
+			if (SetAccessControlProtection(_isProtected, _preserveInheritance))
 			{
-				GetFilePermissions(); // Refresh file permissions
+				// Refresh file permissions
+				GetFilePermissions();
 			}
```
Comment (ferrariofilippo) [inline src/Files.App/ViewModels/Properties/SecurityViewModel.cs:151]: I think comment should be placed outside of the `if`. You can also remove brackets

File: src/Files.App/ViewModels/Properties/SecurityViewModel.cs [hunk 7]
```diff
@@ -139,18 +119,43 @@ public string DisableInheritanceOption
...
 		}
 
 		private void SetDisableInheritanceOption(string options)
 		{
-			isProtected = Boolean.Parse(options.Split(',')[0]);
-			preserveInheritance = Boolean.Parse(options.Split(',')[1]);
+			_isProtected = bool.Parse(options.Split(',')[0]);
+			_preserveInheritance = bool.Parse(options.Split(',')[1]);
+
```
Comment (ferrariofilippo) [inline src/Files.App/ViewModels/Properties/SecurityViewModel.cs:158]: It may be worth to split `options` only once

File: src/Files.App/ViewModels/Properties/SecurityViewModel.cs [hunk 8]
```diff
@@ -195,63 +200,68 @@ private async void EditOwner()
...
+		public bool SetAccessControlProtection(bool isProtected, bool preserveInheritance)
 		{
-			bool isFolder = Item.PrimaryItemAttribute == Windows.Storage.StorageItemTypes.Folder && !Item.IsShortcut;
+			bool isFolder = Item.PrimaryItemAttribute == StorageItemTypes.Folder && !Item.IsShortcut;
 
 			return FileOperationsHelpers.SetAccessRuleProtection(Item.ItemPath, isFolder, isProtected, preserveInheritance);
 		}
+
+		public static Task<string?> OpenObjectPicker()
+			=> FileOperationsHelpers.OpenObjectPickerAsync(NativeWinApiHelper.CoreWindowHandle.ToInt64());
```
Comment (ferrariofilippo) [inline src/Files.App/ViewModels/Properties/SecurityViewModel.cs:264]: Can you remove arrow function?

File: src/Files.App/Filesystem/Security/AccessControlList.cs [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+				fss = null;
+
+				return false;
+			}
+			catch (Exception)
+			{
+				fss = null;
+
+				return false;
+			}
```
Comment (ferrariofilippo) [inline src/Files.App/Filesystem/Security/AccessControlList.cs:251]: You can move `fss = null; return false;` to the end of the method, this way we would avoid code duplication

File: src/Files.App/Filesystem/Security/AccessControlEntryAdvanced.cs [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+		public bool ModifyAccess => AccessMaskFlags.HasFlag(AccessMaskFlags.Modify);
+		public bool FullControlAccess => AccessMaskFlags.HasFlag(AccessMaskFlags.FullControl);
+		public bool SpecialAccess
+			=> (AccessMaskFlags &
+					~AccessMaskFlags.Synchronize &
+					(FullControlAccess ? ~AccessMaskFlags.FullControl : AccessMaskFlags.FullControl) &
+					(ModifyAccess ? ~AccessMaskFlags.Modify : AccessMaskFlags.FullControl) &
+					(ReadAndExecuteAccess ? ~AccessMaskFlags.ReadAndExecute : AccessMaskFlags.FullControl) &
+					(ReadAccess ? ~AccessMaskFlags.Read : AccessMaskFlags.FullControl) &
+					(WriteAccess ? ~AccessMaskFlags.Write : AccessMaskFlags.FullControl)) != 0;
```
Comment (ferrariofilippo) [inline src/Files.App/Filesystem/Security/AccessControlEntryAdvanced.cs:170]: Indent looks strange here. Why are lines 164 and 165 splitted?

**Review by ferrariofilippo — CHANGES_REQUESTED** [review]
I left some comments regarding code style. If I have any time, I'll try to make a second code-review and test this out

**Comment by onein528** [comment]
region-endregion will be removed when I finish. so please ignore them:)

**Comment by ferrariofilippo** [comment]
> So sorry about that 😢

Don't worry

File: src/Files.App/Filesystem/Security/AccessControlEntry.cs [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+				if (item.AccessControlType == AccessControlType.Deny)
+				{
+					if (item.IsInherited)
+					{
+						ace.InheritedDenyAccessMaskFlags |= item.AccessMaskFlags;
+					}
+					else
+					{
+						ace.DeniedAccessMaskFlags |= item.AccessMaskFlags;
+					}
```
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App/Filesystem/Security/AccessControlEntry.cs]: @yair100 Are you sure? brackets of oneline-if-else can be removed?

File: src/Files.App/Helpers/FileOperationsHelpers.cs [hunk 12]
```diff
@@ -629,19 +629,19 @@ public static bool SetLinkIcon(string filePath, string iconFile, int iconIndex)
 			return false;
 		}
 
-		public static FilePermissions GetFilePermissions(string filePath, bool isFolder)
-			=> FilePermissions.FromFilePath(filePath, isFolder);
+		public static AccessControlList GetFilePermissions(string filePath, bool isFolder)
+			=> AccessControlList.FromPath(filePath, isFolder);
```
[the remark this answers is not in the document]
Comment (onein528) [reply src/Files.App/Helpers/FileOperationsHelpers.cs:633]: That's unrelated change. Limited only referactoring App.Filesystem.Security

File: src/Files.App/Filesystem/Security/AccessControlEntryAdvanced.cs (hunk 10 above)
Comment (onein528) [reply src/Files.App/Filesystem/Security/AccessControlEntryAdvanced.cs:170]: Removed one tab each lines.

File: src/Files.App/ViewModels/Properties/SecurityViewModel.cs (hunk 8 above)
Comment (onein528) [reply src/Files.App/ViewModels/Properties/SecurityViewModel.cs:263]: Why? Should be? I didn't change to use arrow, though.

**Comment by onein528** [comment]
The checked items have passed the test. It means that they work correctly.

- [x] Display ACL and all ACEs
- [x] Add an ACE
- [ ] Remove an ACE
- [ ] Change ACE type
- [ ] Change access mask
- [ ] Change inheritance flags

**Comment by onein528** [comment]
I found some issues

- Even though you change ACE's type, Files won't change (there's no functionality)
- Even though you change access mask flags, Files won't change (there's no functionality)
- Even though you change inheritance flags, Files won't change (there's no functionality)

**Comment by onein528** [comment]
@ferrariofilippo We should have corresponding class with classes in System namespace.

- AccessMaskFlags
- InheritanceFlags
- PropagationFlags

File: src/Files.App/Views/Properties/SecurityAdvancedPage.xaml [hunk 13]
```diff
@@ -225,196 +228,198 @@
 							</Border>
 						</ListView.Header>
 						<ListView.ItemTemplate>
-							<DataTemplate x:DataType="perm:FileSystemAccessRuleForUI">
```
Comment (onein528) [inline src/Files.App/Views/Properties/SecurityAdvancedPage.xaml:228]: To bind `IsSelected="{x:Bind IsSelected, Mode=TwoWay}"` in xaml code. This've been changed in the code-beind. No breaking changes.

**Comment by yaira2** [comment]
> Rename Files.App.Filesystem.Permissions with Files.App.Filesystem.Security

I don't think we should make this change, the feature is known internally as file permissions and I think the code should reflect that.

File: src/Files.App/Filesystem/Security/AccessMaskItem.cs [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+			_ace.PropertyChanged += AccessControlEntry_PropertyChanged;
+		}
+
+		public string AccessMaskName { get; set; }
+
+		public AccessMaskFlags AccessMask { get; set; }
+
+		public bool IsEnabled
+		{
+			get => _ace.AccessMaskFlags.HasFlag(AccessMask);
```
Comment (yaira2) [inline src/Files.App/Filesystem/Security/AccessMaskItem.cs:25]: Guideline is to use brackets

**Comment by yaira2** [comment]
> I found some issues
> 
> * Even though you change ACE's type, Files won't change (there's no functionality)
> * Even though you change access mask flags, Files won't change (there's no functionality)
> * Even though you change inheritance flags, Files won't change (there's no functionality)
> 
> Those issues were not created by my refactoring; it means originally existed. So I think I should not fix(I will probably open a PR for those).

Can you open separate issues to track these?

**Comment by onein528** [comment]
Each of the three?

**Comment by yaira2** [comment]
Yes

File: src/Files.App/Filesystem/Security/AccessControlEntry.cs [hunk 15]
```diff
@@ -0,0 +1,612 @@
...
+
+		private List<AccessMaskItem> _accessMaskItemList;
+		public List<AccessMaskItem> AccessMaskItemList
+		{
+			get => _accessMaskItemList;
+			set => SetProperty(ref _accessMaskItemList, value);
+		}
+
+		public RelayCommand<string> ChangeAccessControlTypeCommand { get; set; }
+		public RelayCommand<string> ChangeInheritanceFlagsCommand { get; set; }
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/Security/AccessControlEntry.cs:215]: Would those be any interested to be transformed into RichCommands? Might be overkill, but worth asking.

File: src/Files.App/Filesystem/Security/AccessControlEntry.cs [hunk 16]
```diff
@@ -0,0 +1,612 @@
...
+			//{
+			//	_aceAdvanced.Add(new AccessControlEntryAdvanced(_isFolder)
+			//	{
+			//		AccessControlType = AccessControlType.Deny,
+			//		AccessMaskFlags = DeniedAccessMaskFlags,
+			//		PrincipalSid = Principal.Sid,
+			//		InheritanceFlags = _isFolder ? InheritanceFlags.ContainerInherit | InheritanceFlags.ObjectInherit : InheritanceFlags.None,
+			//		PropagationFlags = PropagationFlags.None
+			//	});
+			//}
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/Security/AccessControlEntry.cs:416]: Why keep this commented code?

File: src/Files.App/Filesystem/Security/AccessControlList.cs [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+				acl.AccessControlEntryPrimitiveMappings.AddRange(rules);
+				acl.AccessControlEntries = new(rules.Select(x => new AccessControlEntry(x, isFolder)));
+
+				return acl;
+			}
+			else
+			{
+				return new(path, isFolder);
+			}
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/Security/AccessControlList.cs:102]: `Else` keyword is redundant here.

File: src/Files.App/Filesystem/Security/AccessControlList.cs [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+					fileSystemSecurity = FileSystemAclExtensions.GetAccessControl(new DirectoryInfo(path));
+					return true;
+				}
+				else if (File.Exists(path))
+				{
+					fileSystemSecurity = FileSystemAclExtensions.GetAccessControl(new FileInfo(path));
+					return true;
+				}
+				// The requested file or folder does not exist
+				else { }
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/Security/AccessControlList.cs:168]: `Else` is redundant here. Kept to have the comment?

File: src/Files.App/Filesystem/Security/AccessMaskFlags.cs [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+
+		/// <summary>
+		/// Create folders
+		/// </summary>
+		CreateDirectories = 4,
+
+		/// <summary>
+		/// Read extended attributes
+		/// </summary>
+		ReadExtendedAttributes = 8,
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/Security/AccessMaskFlags.cs:49]: I'm no expert in security and mask access, but having flags with different meaning sharing the same values seems off... Is this normal?

File: src/Files.App/Filesystem/Security/AccessMaskItem.cs [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+	/// </summary>
+	public class AccessMaskItem : ObservableObject
+	{
+		public AccessMaskItem(AccessControlEntry ace, bool isEditable = true)
+		{
+			IsEditable = isEditable;
+
+			_ace = ace;
+			_ace.PropertyChanged += AccessControlEntry_PropertyChanged;
+		}
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/Security/AccessMaskItem.cs:17]: The constructor should not be that high up in the file. It should be move after properties and fields.

File: src/Files.App/Filesystem/Security/Principal.cs [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+	/// Represents a principal of an ACE or an owner of an ACL.
+	/// </summary>
+	public class Principal : ObservableObject
+	{
+		public Principal()
+		{
+			Groups = new();
+
+			PrincipalType = PrincipalType.Unknown;
+		}
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/Security/Principal.cs:19]: Constructor placement.

File: src/Files.App/ViewModels/Properties/SecurityViewModel.cs [hunk 22]
```diff
@@ -3,255 +3,195 @@
...
-		}
+		private bool _preserveInheritance;
 
-		private void ReplaceChildPermissions()
+		public RelayCommand ChangeOwnerCommand { get; set; }
+		public RelayCommand AddAccessControlEntryCommand { get; set; }
+		public RelayCommand RemoveAccessControlEntryCommand { get; set; }
+		public RelayCommand DisableInheritanceCommand { get; set; }
+		public RelayCommand<string> SetDisableInheritanceOptionCommand { get; set; }
+		public RelayCommand ReplaceChildPermissionsCommand { get; set; }
```
Comment (QuaintMako) [inline src/Files.App/ViewModels/Properties/SecurityViewModel.cs:99]: Rich Commands?

File: src/Files.App/ViewModels/Properties/SecurityViewModel.cs [hunk 23]
```diff
@@ -3,255 +3,195 @@
...
-					// No remaining rules, remove user from list
-					FilePermissions.RulesForUsers.Remove(SelectedRuleForUser);
-				}
-			}
+			OnPropertyChanged(nameof(DisableInheritanceOption));
+			DisableInheritanceCommand.NotifyCanExecuteChanged();
 		}
 
-		public void GetFilePermissions()
+		private void ReplaceChildPermissions()
```
Comment (QuaintMako) [inline src/Files.App/ViewModels/Properties/SecurityViewModel.cs:178]: This method is empty. Why keep it?

File: src/Files.App/Filesystem/Security/AccessMaskFlags.cs (hunk 19 above)
Comment (onein528) [reply src/Files.App/Filesystem/Security/AccessMaskFlags.cs:49]: Yes. You can refer to ACCESS_MASK in MSDN.

File: src/Files.App/ViewModels/Properties/SecurityViewModel.cs (hunk 23 above)
Comment (onein528) [reply src/Files.App/ViewModels/Properties/SecurityViewModel.cs:178]: This had already been empty before having worked on refactoring.

Comment (onein528) [reply src/Files.App/ViewModels/Properties/SecurityViewModel.cs:178]: I can add this. But we should limit the scope of this PR.

File: src/Files.App/Filesystem/Security/AccessControlList.cs (hunk 17 above)
Comment (onein528) [reply src/Files.App/Filesystem/Security/AccessControlList.cs:102]: Not redundant.

Comment (QuaintMako) [reply src/Files.App/Filesystem/Security/AccessControlList.cs:102]: It actually is, there is a `return` inside the `if`, and no other instructions after the `else`.
But it's really small a detail.

**Review by QuaintMako — COMMENTED** [review]
I won't give my approval for the PR because I am not versed enough in the security to be sure at 100%, but alongside the two other reviews, I believe it's good!

Thanks for the huge refactoring and cleaning @onein528 ! (And for the others you've done as well)

**Review by QuaintMako — APPROVED** [review]

File: src/Files.App/Filesystem/Security/InheritanceFlags.cs [hunk 24]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
+
+namespace Files.App.Filesystem.Security
+{
+	/// <summary>
+	/// Represents inheritance flags of an ACE
+	/// </summary>
+	[Flags]
+	public enum InheritanceFlags
+	{
```
[the remark this answers is not in the document]
Comment (yaira2) [reply src/Files.App/Filesystem/Security/InheritanceFlags.cs:10]: @onein528

[the remark this answers is not in the document]
Comment (onein528) [reply src/Files.App/Filesystem/Security/InheritanceFlags.cs:10]: There's same name enum class in System.Security. Should be same.

File: src/Files.App/Filesystem/Security/PropagationFlags.cs [hunk 25]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
+
+namespace Files.App.Filesystem.Security
+{
+	/// <summary>
+	/// Represents propagation flags of an ACE
+	/// </summary>
+	[Flags]
+	public enum PropagationFlags
+	{
```
[the remark this answers is not in the document]
Comment (onein528) [reply src/Files.App/Filesystem/Security/PropagationFlags.cs:10]: There's same name enum class in System.Security. Should be same.

File: src/Files.App/Filesystem/Security/AccessMaskFlags.cs [hunk 26]
```diff
@@ -0,0 +1,131 @@
+﻿using System;
+
+namespace Files.App.Filesystem.Security
+{
+	/// <summary>
+	/// Represents access mask flags of an ACE.
+	/// </summary>
+	[Flags]
+	public enum AccessMaskFlags
+	{
```
[the remark this answers is not in the document]
Comment (onein528) [reply src/Files.App/Filesystem/Security/AccessMaskFlags.cs:10]: There's same name enum class in System.Security. Should be same.

File: src/Files.App/Filesystem/Security/InheritanceFlags.cs (hunk 24 above)
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App/Filesystem/Security/InheritanceFlags.cs:10]: @yair100
